sdks/python/apache_beam/testing/benchmarks/nexmark/queries/query3.py

```diff
@@ -0,0 +1,158 @@
+                  ResultNames.CITY: t[1].city,
+                  ResultNames.STATE: t[1].state,
+                  ResultNames.AUCTION_ID: t[0].id
+              }))
+
+
+class JoinFn(beam.DoFn):
+  """
+  Join auctions and person by person id and emit their product one pair at
+  a time.
+
+  We know a person may submit any number of auctions. Thus new person event
+  must have the person record stored in persistent state in order to match
+  future auctions by that person.
+
+  However we know that each auction is associated with at most one person, so
+  only need to store auction records in persistent state until we have seen the
+  corresponding person record. And of course may have already seen that record.
+  """
+
+  AUCTIONS = 'auctions_state'
+  PERSON = 'person_state'
+  PERSON_EXPIRING = 'person_state_expiring'
+
+  auction_spec = userstate.BagStateSpec(AUCTIONS, nexmark_model.Auction.CODER)
+  person_spec = userstate.ReadModifyWriteStateSpec(
+      PERSON, nexmark_model.Person.CODER)
+  person_timer_spec = userstate.TimerSpec(
+      PERSON_EXPIRING, userstate.TimeDomain.WATERMARK)
+
+  def __init__(self, max_auction_wait_time):
+    self.max_auction_wait_time = max_auction_wait_time
+
+  def process(
+      self,
+      element,
+      auction_state=beam.DoFn.StateParam(auction_spec),
+      person_state=beam.DoFn.StateParam(person_spec),
+      person_timer=beam.DoFn.TimerParam(person_timer_spec)):
+    # extract group with tags from element tuple
+    _, group = element
+
+    existing_person = person_state.read()
+    if existing_person:
+      # the person exists in person_state for this person id
+      for auction in group[nexmark_query_util.AUCTION_TAG]:
+        yield auction, existing_person
+      return
+
+    new_person = None
+    for person in group[nexmark_query_util.PERSON_TAG]:
+      if not new_person:
+        new_person = person
+      else:
+        logging.error(
+            'two new person wtih same key: %s and %s' % (person, new_person))
+        continue
+      # read all pending auctions for this person id, output and flush it
+      pending_auctions = auction_state.read()
+      if pending_auctions:
+        for pending_auction in pending_auctions:
+          yield pending_auction, new_person
+        auction_state.clear()
+      # output new auction for this person id
+      for auction in group[nexmark_query_util.AUCTION_TAG]:
+        yield auction, new_person
+      # remember person for max_auction_wait_time seconds for future auctions
+      person_state.write(new_person)
+      person_timer.set(new_person.date_time + self.max_auction_wait_time)
+    # we are done if we have seen a new person
+    if new_person:
+      return
+
+    # remember auction until we see person
+    for auction in group[nexmark_query_util.AUCTION_TAG]:
+      auction_state.add(auction)
+
+  @on_timer(person_timer_spec)
+  def expiry(self, person_state=beam.DoFn.StateParam(person_spec)):
+    person_state.clear()
```

Comment:
Can it happen that the same person creates new auctions after > `max_auction_waiting_time` ? Will there be a new person event? If we get new auctions after the person is expired, then we'll just keep adding them to auction_state forevetr, no?